pkg/occlient/occlient.go

```diff
@@ -737,30 +933,85 @@ func (c *Client) StartBinaryBuild(name string, path string, asFile bool) error {
 	return nil
 }
 
-// StartBuild starts new build as it is
-func (c *Client) StartBuild(name string) error {
+// StartBuild starts new build as it is, returns name of the build stat was started
+func (c *Client) StartBuild(name string) (string, error) {
 	buildRequest := buildv1.BuildRequest{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: name,
 		},
 	}
 	result, err := c.buildClient.BuildConfigs(c.namespace).Instantiate(name, &buildRequest)
 	if err != nil {
-		return errors.Wrapf(err, "unable to start build %s", name)
+		return "", errors.Wrapf(err, "unable to start build %s", name)
 	}
-	log.Debugf("Build %s triggered.", name)
+	log.Debugf("Build %s for BuildConfig %s triggered.", name, result.Name)
 
-	err = c.FollowBuildLog(result.Name)
+	return result.Name, nil
+}
+
+// WaitForBuildToFinish block and waits for build to finish. Returns error if build failed or was canceled.
+func (c *Client) WaitForBuildToFinish(buildName string) error {
+	log.Debugf("Waiting for %s  build to finish", buildName)
+
+	w, err := c.buildClient.Builds(c.namespace).Watch(metav1.ListOptions{
+		FieldSelector: fields.Set{"metadata.name": buildName}.AsSelector().String(),
+	})
 	if err != nil {
-		return errors.Wrapf(err, "unable to start build %s", name)
+		return errors.Wrapf(err, "unable to watch build")
+	}
+	defer w.Stop()
+	for {
+		val, ok := <-w.ResultChan()
+		if !ok {
+			break
+		}
+		if e, ok := val.Object.(*buildv1.Build); ok {
+			log.Debugf("Status of %s build is %s", e.Name, e.Status.Phase)
+			switch e.Status.Phase {
+			case buildv1.BuildPhaseComplete:
+				log.Debugf("Build %s completed.", e.Name)
+				return nil
+			case buildv1.BuildPhaseFailed, buildv1.BuildPhaseCancelled, buildv1.BuildPhaseError:
+				return errors.Errorf("build %s status %s", e.Name, e.Status.Phase)
+			}
+		}
```

Comment:
No timeout?

Reply:
So for BuildConfig there is an optional field `completionDeadlineSeconds` that determines the maximal duration of a build. We should probably rely on OpenShift Failing build after the `completionDeadlineSeconds`.  We can't add some arbitrary timeout here, that will be different from what we or user sets in `completionDeadlineSeconds`.
OpenShift will kill build after `completionDeadlineSeconds` If that happens it will result `buildv1.BuildPhaseFailed` state and that is handled.